sling/nlp/embedding/word-embeddings.cc

```diff
@@ -0,0 +1,517 @@
+    vocabulary_.Load(task->GetInputFile("vocabulary"), subsampling_);
+    int vocabulary_size = vocabulary_.size();
+
+    // Build embedding model.
+    flow_.inputs = flow_.outputs = vocabulary_size;
+    flow_.dims = embedding_dims_;
+    flow_.in_features = window_ * 2;
+    flow_.Build();
+
+    // Compile embedding model.
+    myelin::Compiler compiler;
+    myelin::Network model;
+    compiler.Compile(&flow_, &model);
+
+    // Initialize weights.
+    Random rnd;
+    myelin::TensorData W0 = model[flow_.W0];
+    myelin::TensorData W1 = model[flow_.W1];
+    for (int i = 0; i < vocabulary_size; ++i) {
+      for (int j = 0; j < embedding_dims_; ++j) {
+        W0.at<float>(i, j) = rnd.UniformFloat(1.0, -0.5);
+        W1.at<float>(i, j) = 0.0;
+      }
+    }
+
+    // Initialize commons store.
+    commons_ = new Store();
+    docnames_ = new DocumentNames(commons_);
+    commons_->Freeze();
+
+    // Statistics.
+    num_documents_ = task->GetCounter("num_documents");
+    total_documents_ = task->GetCounter("total_documents");
+    num_tokens_ = task->GetCounter("num_tokens");
+    num_instances_ = task->GetCounter("num_instances");
+    epochs_completed_ = task->GetCounter("epochs_completed");
+
+    // Start training threads. Use one worker thread per input file.
+    std::vector<string> filenames = task->GetInputFiles("documents");
+    int threads = filenames.size();
+    task->Fetch("threads", &threads);
+    WorkerPool pool;
+    pool.Start(threads, [this, &filenames, &model](int index) {
+      Worker(index, filenames[index % filenames.size()], &model);
+    });
+
+    // Wait until workers completes.
+    pool.Join();
+
+    // Output profile.
+    myelin::LogProfile(model);
+
+    // Write embeddings to output file.
+    const string &output_filename = task->GetOutputFile("output");
+    EmbeddingWriter writer(output_filename, vocabulary_size, embedding_dims_);
+    std::vector<float> embedding(embedding_dims_);
+    for (int i = 0; i < vocabulary_size; ++i) {
+      const string &word = vocabulary_.word(i);
+      for (int j = 0; j < embedding_dims_; ++j) {
+        embedding[j] = W1.at<float>(i, j);
+      }
+      writer.Write(word, embedding);
+    }
+    CHECK(writer.Close());
+
+    // Clean up.
+    vocabulary_.Clear();
+    docnames_->Release();
+    delete commons_;
+    docnames_ = nullptr;
+    commons_ = nullptr;
+  }
+
+  // Worker thread for training embedding model.
+  void Worker(int index, const string &filename, myelin::Network *model) {
+    Random rnd;
+    rnd.seed(index);
+    int epoch = 0;
+    std::vector<int> words;
+
+    // Set up model compute instances.
+    myelin::Instance l0(model->GetCell(flow_.layer0));
+    myelin::Instance l1(model->GetCell(flow_.layer1));
+    myelin::Instance l0b(model->GetCell(flow_.layer0b));
+
+    int *features = l0.Get<int>(model->GetParameter(flow_.fv));
+    int *fend = features + 2 * window_;
```

Comment:
flow_.in_features instead of 2 * window_ ?

Reply:
fixed